feat: Digitization 2, introduce SurfaceMasker

asalzburger opened the pull request:
This PR is the second step of the new Digitisation.

It introduces the `SurfaceMasker` that is run before the `Channelizer` which was introduced in part 1.
The `PlanarSurfaceMasker` clips/masks a segment onto the surface (= readout) bounds.

Additionally, it harmonizes the input/output between Channelized / Masker.

It also reorders the PlanarSurface testbeds to be used for `ChannelizerTest` and `PlanarSurfaceMaskTest`.

* `RectangleBounds`
- green : inside from the the start (untouched)
- blue: clipped

![Unknown-2](https://user-images.githubusercontent.com/26623879/97012241-f576d880-1547-11eb-93a5-9325e6475821.png)

* `TrapezoidBounds`

![Unknown](https://user-images.githubusercontent.com/26623879/97012362-13443d80-1548-11eb-8332-203401fda2d3.png)

* `RadialBounds`

- ref: segment fully outside

![Unknown-3](https://user-images.githubusercontent.com/26623879/97013370-4509d400-1549-11eb-9bba-dfdd78d61ce4.png)

asalzburger commented:
Hi @XiaocongAi - this should answer your question about outside segments.

XiaocongAi commented on Tests/UnitTests/Fatras/Digitization/PlanarSurfaceMaskTests.cpp:
```diff
@@ -0,0 +1,197 @@
+  CHECK_CLOSE_ABS(Acts::VectorHelpers::phi(clipped[0]), M_PI_4,
+                  Acts::s_epsilon);
+
+  /// Case four: outside phi max
```
```suggestion
  /// Case five: outside phi max
```

asalzburger replied:
Solved.

XiaocongAi reviewed (commented):
Hi @asalzburger , it looks good to me. There is only the Annulus edge vector I'm not very sure (But since there is the unit test for this, I guess you should be right).

asalzburger commented on Fatras/include/ActsFatras/Digitization/PlanarSurfaceMask.hpp:
```diff
@@ -0,0 +1,63 @@
+  ///
+  /// @note Only PlaneSurface/DiscSurface are supported
+  ///
+  /// @param geoCtx The geometry context
```
Fixed.

asalzburger commented on Fatras/include/ActsFatras/Digitization/PlanarSurfaceMask.hpp:
```diff
@@ -0,0 +1,63 @@
+  /// @param inside The inside point of the segment
+  /// @param vertices The vertices of the polygon
+  ///
+  /// @return a result wrapping the new new outside position
```
Fixed.

asalzburger commented on Fatras/include/ActsFatras/Digitization/PlanarSurfaceMask.hpp:
```diff
@@ -0,0 +1,63 @@
+
+  /// Apply the mask of a polygon
+  ///
+  /// @param outise The outside point of the segment
```
Fixed.

asalzburger commented on Fatras/src/Digitization/DigitizationError.cpp:
```diff
@@ -20,8 +20,10 @@ std::string ActsFatras::detail::DigitizationErrorCategory::message(
+    case DigitizationError::UndefinedSurface:
       return "Digitization: no surface for bound measurement defined.";
+    case DigitizationError::MaskingError:
+      return "Digitization: surface maks could not be applied.";
```
Fixed.

asalzburger commented on Tests/UnitTests/Fatras/Digitization/DigitizationCsvOutput.hpp:
```diff
@@ -0,0 +1,55 @@
+  /// @param outf The output stream
+  /// @param r The radius of the arc
+  /// @param phiMin The minimum phi
+  /// @param phiMin The maximum phi phi
```
Fixed/

asalzburger commented on Fatras/src/Digitization/PlanarSurfaceMask.cpp:
```diff
@@ -0,0 +1,232 @@
+          [](const Acts::Vector2D& a, const Acts::Vector2D& b) {
+            return (Acts::VectorHelpers::phi(a) < Acts::VectorHelpers::phi(b));
+          });
+      Segment2D minEdge = {cartesianEdges[0], cartesianEdges[3]};
```
I will check.

asalzburger replied:
I rewrote that part.

asalzburger commented:
There was a problem/bug in the `DiscTrapezoidBounds`, which was only partly fixed in the `Digitization 1` PR, this fixes it completely now.

asalzburger commented:
![Unknown](https://user-images.githubusercontent.com/26623879/97173654-36156280-1791-11eb-8ff8-ca0028c384e8.png)

This is for `DiscTrapezoidBounds`.

asalzburger commented:
Still investigating the `AnnulusBounds` issue, so changing to `WIP` now.

asalzburger commented on Fatras/src/Digitization/PlanarSurfaceMask.cpp:
```diff
@@ -0,0 +1,232 @@
+      Segment2D maxEdge = {cartesianEdges[1], cartesianEdges[2]};
+      sectorEdges = {minEdge, maxEdge};
+      std::pair<double, double> phisMinR = {
+          Acts::VectorHelpers::phi(cartesianEdges[0]),
```
Again, I rewrote that part.

asalzburger commented:
![Unknown-1](https://user-images.githubusercontent.com/26623879/97598300-16876f80-1a07-11eb-8802-c1ec73d1af19.png)

Now with fixed unit test & bug fix in AnnulusBounds.

asalzburger commented on Core/include/Acts/Surfaces/AnnulusBounds.hpp:
```diff
@@ -154,7 +154,6 @@ class AnnulusBounds : public DiscBounds {
   Vector2D m_moduleOrigin;
   Vector2D m_shiftXY;  // == -m_moduleOrigin
   Vector2D m_shiftPC;
-  double m_phiAvg;
```
This was the bug mentioned, we used undefined `m_phiAvg` which resulted in useless results.

asalzburger commented:
Hi @XiaocongAi  - this is now updated, I shuffled the code around, but it's largely the same. Also, I fixed the problem with the `AnnulusBounds` which had an internal bug.

XiaocongAi commented on Core/src/Surfaces/AnnulusBounds.cpp:
```diff
@@ -100,41 +99,47 @@ std::vector<Acts::Vector2D> Acts::AnnulusBounds::corners() const {
+    std::vector<Acts::Vector2D> rvertices;
+
+    // Shift them to get the module phi
+    double phiMinInner = VectorHelpers::phi(m_inLeftStripXY - m_moduleOrigin);
```
Could it be that `phiMinInner` is derived from `m_inRightStripXY` and `phiMaxInner` derived from `m_inLeftStripXY`?

asalzburger replied:
Left, right - that's so complicated :-P

I will fix this.

asalzburger replied:
Done.

XiaocongAi reviewed (commented):
Hi @asalzburger , the updated approach to get the segment intersection looks more clear. But there are some points I'm not so sure.

XiaocongAi commented on Core/src/Surfaces/AnnulusBounds.cpp:
```diff
@@ -100,41 +99,47 @@ std::vector<Acts::Vector2D> Acts::AnnulusBounds::corners() const {
+    std::vector<double> phisOuter =
+        detail::VerticesHelper::phiSegments(phiMinOuter, phiMaxOuter);
+
+    // Inner bow from phi_min -> phi_max
```
Since it's starting from the left inner corner, it's `phi_max->phi_min`?

asalzburger replied:
Will correct with left/right correctly.

asalzburger replied:
It actually worked by accident (at least for the obj displayer), because it has to create a full closed face.
Fixed now.

asalzburger commented on Fatras/src/Digitization/PlanarSurfaceMask.cpp:
```diff
@@ -0,0 +1,276 @@
+    if (sR < radii[0] or eR < radii[0]) {
+      intersectCircle(radii[0]);
+    }
+    if (sR > radii[1] or eR > radii[1]) {
```
Hi @XiaocongAi - you are good. Indeed, this is the case that could happen, BUT ... This masker will operate on hits that are on surface and the reason why they could be outside of the mask is purely due to the incident angle and the charge drift. 

Both makes it practically impossible to create a case you describe.

As > 99 % of all cases are safe with both inside and the masker will not even operate, I do not want to penalise the 99% good cases with extra checks, that are pathological.

I will clarify this in the description.

XiaocongAi replied:
Yeah, from the realistic point of view, we are indeed safe.

asalzburger replied:
This is what I added:

```c++
  /// @note If both end points of the segment are inside, the segment
  /// is not clipped/masked, even if it would cross a surface boundary.
  /// Examples for those would be non-covex polygons or segments on a
  /// radial bound, where the radial boundary is crossed. Such segments
  /// do not occur in Digitization, as the hit has to be inside the 
  /// surface bounds to start with.
```

XiaocongAi replied:
Great! :+1:

asalzburger commented on Fatras/src/Digitization/PlanarSurfaceMask.cpp:
```diff
@@ -0,0 +1,276 @@
+
+/// Helper method to apply the mask and return.
+///
+/// If two (ore more) intersections would be good, appply the first two
```
Fixed.

asalzburger commented on Fatras/src/Digitization/PlanarSurfaceMask.cpp:
```diff
@@ -0,0 +1,276 @@
+///
+/// If two (ore more) intersections would be good, appply the first two
+/// If only one is available, the boolean tells you which one it is.
+/// If no intersection is valide, return an error code for masking.
```
Fixed.

asalzburger commented on Fatras/src/Digitization/PlanarSurfaceMask.cpp:
```diff
@@ -0,0 +1,276 @@
+                        sLength);
+    }
+    if (sR > radii[1] or eR > radii[1]) {
+      checkIntersection(intersections,
```
I know, the idea is the following:
- if any of them it outside : create a 2D intersection (which one is found out later).

asalzburger reviewed (commented):
Addressed all comments.

XiaocongAi reviewed (approved):
Thanks! Approve.

asalzburger commented:
Waiting for CI to complete before merging.
